### PR TITLE
Issue #13693: migrate header modules to use property macro

### PR DESCRIPTION
**Pull request by romani**
Issue #13693

all extra refactoring was enforced by our verify validation.

setCharset
https://github.com/checkstyle/checkstyle/commit/326e3ae67b8cc2a1dcbcf2356b30fd9a4afa964a#diff-fa56dd39f138d19b44a0aa110150c621c7ba641aead7a198d35db46fdbc767a2R71

setHeaderFile
https://github.com/checkstyle/checkstyle/commit/6f1f7bd3fdcca84404e97964b3e48de6db5024ec#diff-fa56dd39f138d19b44a0aa110150c621c7ba641aead7a198d35db46fdbc767a2R69

setHeader
https://github.com/checkstyle/checkstyle/commit/6f1f7bd3fdcca84404e97964b3e48de6db5024ec#diff-fa56dd39f138d19b44a0aa110150c621c7ba641aead7a198d35db46fdbc767a2R81

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java:53]: Not expected , but I will improve after merge of blocker PR.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java:53]: How is it not expected and how will it be improved?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java:53]: By doing https://github.com/checkstyle/checkstyle/pull/13964#discussion_r1399822130

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java:53]: How will going against the setter improve the default value? There is no default value in the setter.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java:53]: I have idea to:
If field is not found, default is null for sure. We have scalar types and array. 
For scaral type null is null, for array it will be `{}`

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java]: fixed.

**Comment by romani** [comment]
GitHub, generate website

**Comment by romani** [comment]
GitHub, generate website

**Comment by romani** [comment]
Travis is crashing in JVM, restarting usually helps, I sent mail to support to look if it is related to some infra update on their side.

File: src/xdocs/checks/header/header.xml
Comment (rnveach) [inline src/xdocs/checks/header/header.xml:65]: Version number changes are not expected. Please show proof 5.0 is wrong.

File: src/xdocs/checks/header/header.xml
Comment (rnveach) [inline src/xdocs/checks/header/header.xml:79]: Same.

File: src/xdocs/checks/header/header.xml
Comment (rnveach) [inline src/xdocs/checks/header/header.xml:86]: Same.

File: src/xdocs/checks/header/regexpheader.xml
Comment (rnveach) [inline src/xdocs/checks/header/regexpheader.xml:104]: Same.

File: src/xdocs/checks/header/regexpheader.xml
Comment (rnveach) [inline src/xdocs/checks/header/regexpheader.xml:125]: Same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java:83]: Also need commits showing this version number is correct.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java:94]: Same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java:158]: Same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheck.java:71]: This is not expected. Why is this changed?

If it is because Javadoc has no examples, this is still copied to xdoc which does.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/xdocs/checks/header/header.xml
Comment (romani) [reply src/xdocs/checks/header/header.xml:65]: Already in PR description

Comment (rnveach) [reply src/xdocs/checks/header/header.xml:65]: Yes, PR description shows original version is correct. Please ensure you are showing before and after on release commits.

> https://github.com/checkstyle/checkstyle/commit/326e3ae67b8cc2a1dcbcf2356b30fd9a4afa964a#diff-fa56dd39f138d19b44a0aa110150c621c7ba641aead7a198d35db46fdbc767a2R71

https://github.com/checkstyle/checkstyle/blob/326e3ae67b8cc2a1dcbcf2356b30fd9a4afa964a/pom.xml#L5
Version 5.0 .

File: config/checkstyle-non-main-files-suppressions.xml
Comment (romani) [reply config/checkstyle-non-main-files-suppressions.xml:254]: fixed

File: src/xdocs/checks/header/header.xml
Comment (romani) [reply src/xdocs/checks/header/header.xml]: all since are reverted. map is updated with values from xdoc

File: src/xdocs/checks/header/header.xml
Comment (romani) [reply src/xdocs/checks/header/header.xml]: all since are reverted. map is updated with values from xdoc

File: src/xdocs/checks/header/header.xml
Comment (romani) [reply src/xdocs/checks/header/header.xml]: all since are reverted. map is updated with values from xdoc

File: src/xdocs/checks/header/regexpheader.xml
Comment (romani) [reply src/xdocs/checks/header/regexpheader.xml]: all since are reverted. map is updated with values from xdoc

File: src/xdocs/checks/header/regexpheader.xml
Comment (romani) [reply src/xdocs/checks/header/regexpheader.xml]: all since are reverted. map is updated with values from xdoc

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: all since are reverted. map is updated with values from xdoc

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: all since are reverted. map is updated with values from xdoc

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: all since are reverted. map is updated with values from xdoc

**Comment by romani** [comment]
> The point of macros is to document everything in our Java files to be pulled over for XDoc documentation.

Yes, but ....
During implementation phase we too much narrowed our minds to xdoc. And we blindly thought that `@since` is also going to serve for xdoc. But I started such PRs you become confused, as you brain is clear xdoc project, and you brain properly interpret `@since` as tag that defines when method is appeared in particular class - https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#@since .
And here there are tiny nuances:
0) Xdoc needs see only properties in result object (that is combination of multiple layers(abstract/parent classes)) 
1) We did not accounted that CS modules has a lot of abstract classes, definitely our big miss.
2) We started to use javadoc general `@since` and it worked well until we came to our old Checks and all this activity is started, and confusion is growing.

Example why general javadoc `@since` is not going to work for us:
Imagine we have check:
```
/** @since 13.0.0 */
class SomeRuleCheck extends AbstractCheck {
...
/** @since 13.0.0 */
setProp1(){}
/** @since 13.0.0 */
setProp2(){}
...
}  
```
we decided to make refactoring in classes to create some abstract class (user and xdoc should not notice anything):
```
/** @since 13.1.1 */
class SomeAbstractCheck extends AbstractCheck {
...
/** @since 13.1.1 */
setProp1(){}
...
} 

/** @since 13.0.0 */
class SomeRuleCheck extends SomeAbstractCheck {
...
/** @since 13.0.0 */
setProp2(){}
...
}  
```

as you can see setProp1 changed since from 13.0.0 to 13.1.1. from Java (javadoc of class) perspective and engineers - correct, from user perspective  someRuleCheckInstance.setProp1 works on cs version 13.0.0.
WE SHOULD NOT reuse general `@since` for xdoc, it will not works clearly as it hard to serve two purposes in same time.
We need `@xdoc-since` tag that will have a value that will serve single purpose - version when such property is accessible by user of object/instance (NOT a some class).

but here is another problem, it is not easy, I failed multiple times to find proper value even for `@xdoc-since` (`@object-since`), at least do not agree with me on what I propose, so I try to avoid this very complicated process and move forward with workaround we already have, and deal with elimination of such hardcoded map in separate PR when we will have nuances gathered and we can make better decision on what to do.

> I have laid out a few times how the logic should work. If there is still confusion on it, I can give it again.

Example that I showed above clearly shows that it will be error-prone for xdoc.

> One of my goals is to minimize this list based on how since values should be documented.

yes, but lets sacrifice (gambit) this for a bit, finish migration of properties, resolve all such problems with unsynced descriptions and typos and ...... and remove copy-paste in templates to unblock https://github.com/checkstyle/checkstyle/pull/13670 and deal with imperfection of hardcoded list for modules with abstract classes. That we are not going to refactor and it is our legacy. 
Something tells me that after we finish we might encounter more problems, that might change our thinking again.
Main point is to deal with version and only versions in separate PRs, so easy to review and focus on nuances and debate what version is actually correct 4.4 or 5.0.

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:184]: These are hard to read, please just suppress multiple string literal check for this file or add some underscores like `VERSION_3_2` and `VERSION_8_24`

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:200]: Translation of `HeaderCheck` to upper snake case is `HEADER_CHECK`, please do for all.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
We need to move forward with migration to templates, I am OK with some hacks that are contained in SiteUtil/etc, with the understanding that we will address these in the future.

Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: done

**Review by nrmancuso — APPROVED** [review]
I am good, let's move forward with this migration.

**Comment by romani** [comment]
@rnveach , please approve or recall review to let me proceed with migration. We will come back to version discussion a bit later.

**Comment by romani** [comment]
I don't see new messages, so I consider reassignment back to us as sign to be skipped from review.

We are moving forward, we will be back to versions after migration

**Pull request merged**
